### PR TITLE
Implement Process.executable_path by using Win32 API

### PR DESCRIPTION
**Pull request by kubo**
This patch uses Win32 API to get the executable path of the current process.

### CODE REVIEW

File: src/process/executable_path.cr [hunk 1]
```diff
@@ -90,6 +90,25 @@ end
       "/proc/self/exe"
     end
   end
+{% elsif flag?(:win32) %}
+  require "crystal/system/windows"
```
Comment (straight-shoota) [inline src/process/executable_path.cr:94]: Why is this necessary?

**Comment by asterite** [comment]
Thank you!

Could you change these `pending_win32` to `it` in this same PR?

https://github.com/crystal-lang/crystal/blob/2d41d70f13a55d0fe2a171edfa2c1f076a129f74/spec/std/process_spec.cr#L305-L322

**Comment by kubo** [comment]
> Could you change these `pending_win32` to `it` in this same PR?
> 
> https://github.com/crystal-lang/crystal/blob/2d41d70f13a55d0fe2a171edfa2c1f076a129f74/spec/std/process_spec.cr#L305-L322

This PR is for `Process.executable_path`. The specs are for `Process.find_executable`. It isn't reasonable to include unrelated spec in a same PR. In addition only one of the three pending tests passed.

**Comment by asterite** [comment]
Oops, sorry. Nevermind then.

**Comment by straight-shoota** [comment]
Having a proper spec for `executable_path` would be nice, but I don't think that's in scope for this PR.

File: src/process/executable_path.cr [hunk 2]
```diff
@@ -90,6 +90,25 @@ end
...
+  require "c/libloaderapi"
+
+  class Process
+    private def self.executable_path_impl
+      hMod = LibC.GetModuleHandleW(nil)
+      Crystal::System.retry_wstr_buffer do |buffer, small_buf|
+        len = LibC.GetModuleFileNameW(hMod, buffer, buffer.size)
```
Comment (asterite) [inline src/process/executable_path.cr:101]: From the [docs](https://docs.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-getmodulefilenamew):

> hModule: A handle to the loaded module whose path is being requested. If this parameter is NULL, GetModuleFileName retrieves the path of the executable file of the current process.

So maybe we don't need to call `GetModuleHandleW`?

Comment (kubo) [reply src/process/executable_path.cr:101]: Thanks for pointing. I'll remove `GetModuleHandleW`.

**Comment by oprypin** [comment]
Umm many of these Win32 things are already implemented, maybe you should hold off on separate implementations

File: src/process/executable_path.cr (hunk 2 above)
Comment (kubo) [reply src/process/executable_path.cr]: I tested it on my local machine and pushed a new commit (568e1ad).

**Comment by oprypin** [comment]
Yes, I think the implementation in https://github.com/crystal-lang/crystal/pull/8518/files#diff-5f66ba2dbcb4e101da5364aa2808471b is better

**Comment by oprypin** [comment]
Ah no, this one is likely better, I confused them :sweat_smile:

File: src/process/executable_path.cr [hunk 3]
```diff
@@ -90,6 +90,24 @@ end
...
+  class Process
+    private def self.executable_path_impl
+      Crystal::System.retry_wstr_buffer do |buffer, small_buf|
+        len = LibC.GetModuleFileNameW(nil, buffer, buffer.size)
+        if 0 < len < buffer.size
+          break String.from_utf16(buffer[0, len])
+        elsif small_buf && len == buffer.size
+          next 32767 # big enough. 32767 is the maximum total path length of UNC path.
+        else
+          raise ::File::Error.from_winerror("Error getting executable path", file: "?")
```
Comment (oprypin) [inline src/process/executable_path.cr:106]: Wait, what does `file: "?"` mean?

Comment (kubo) [reply src/process/executable_path.cr:106]: I should have heard opinions about it first.

At first I wrote the line without `file: "?"` and got the following error.
```
In src/file/error.cr:11:27

 11 | File::NotFoundError.new(message, **opts)
                          ^--
Error: missing argument: file
```
I came up with two ideas. One is using another error class. The other is supplying a dummy file name such as `"?"`. I thought that the error was related to file so the `File::Error` was proper. So I chose the latter. However I have not been satisfied with it yet.

Any idea in order not to use `file: "?"`?

Comment (oprypin) [reply src/process/executable_path.cr:106]: Hm well why is is `::File::Error` in the first place? That is what requires a `file` arg.

BTW I don't see any other `executable_path_impl` raising any errors, they just return `nil`. Maybe it should be that?

If not, maybe go back to `RuntimeError` and possibly specialize the exception in a different direction?

Comment (straight-shoota) [reply src/process/executable_path.cr:106]: It should just return `nil` if the file can't be found, as per https://crystal-lang.org/api/0.34.0/Process.html#executable_path-class-method

Comment (kubo) [reply src/process/executable_path.cr]: @straight-shoota 
Thanks. I fixed it just to return `nil` on error.
